cloud/pkg/upgradecontroller/controller/downstream.go

```diff
@@ -0,0 +1,381 @@
+				klog.Errorf("Failed to get node(%s) info: %v", node, err)
+				continue
+			}
+			if filterVersion(nodeInfo.Status.NodeInfo.KubeletVersion, upgrade.Spec.Version) {
+				klog.Warningf("Node(%s) version(%s) already on the expected version %s.", node, nodeInfo.Status.NodeInfo.KubeletVersion, upgrade.Spec.Version)
+				continue
+			}
+
+			// we only care about edge nodes, so just remove not edge nodes
+			if !isEdgeNode(nodeInfo) {
+				klog.Warningf("Node(%s) is not edge node, just skip", nodeInfo.Name)
+				continue
+			}
+
+			nodeName = append(nodeName, node)
+		}
+	} else if upgrade.Spec.LabelSelector != nil {
+		selector, err := metav1.LabelSelectorAsSelector(upgrade.Spec.LabelSelector)
+		if err != nil {
+			klog.Errorf("LabelSelector(%s) is not valid: %v", upgrade.Spec.LabelSelector, err)
+			return
+		}
+		nodes, err := dc.kubeClient.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{LabelSelector: selector.String()})
+		if err != nil {
+			klog.Errorf("Failed to get nodes with label %s: %v", selector.String(), err)
+			return
+		}
+
+		for _, node := range nodes.Items {
+			if filterVersion(node.Status.NodeInfo.KubeletVersion, upgrade.Spec.Version) {
+				klog.Warningf("Node(%s) version(%s) already on the expected version %s.", node, node.Status.NodeInfo.KubeletVersion, upgrade.Spec.Version)
+				continue
+			}
+
+			// we only care about edge nodes, so just remove not edge nodes
+			if !isEdgeNode(&node) {
+				klog.Warningf("Node(%s) is not edge node, just skip", node.Name)
+				continue
+			}
+
+			nodeName = append(nodeName, node.Name)
+		}
+	} else {
+		klog.Errorf("Upgrade prohibit, no valid nodes are specified: no nodeNames and labelSelector are specified.")
+		return
+	}
+
+	// deduplicate: remove duplicate nodes to avoid repeating upgrade to the same node
+	nodeName = RemoveDuplicateElement(nodeName)
+
+	klog.Infof("Filtered finished, the below nodes are to upgrade\n%v\n", nodeName)
+
+	for _, node := range nodeName {
+		// send upgrade msg to every edge node
+		msg := model.NewMessage("")
+
+		resource := buildUpgradeResource(upgrade.Name, node)
+
+		// get HistoryID
+		var newHistoryID int
+		for index := range upgrade.Status.UpgradeStatus {
+			if upgrade.Status.UpgradeStatus[index].NodeName == node {
+				// historyID is 0,1,2...
+				newHistoryID = len(upgrade.Status.UpgradeStatus[index].History)
+				break
+			}
+		}
+		upgradeReq := commontypes.UpgradeRequest{
+			UpgradeID:       upgrade.Name,
+			UpgradeTool:     upgrade.Spec.UpgradeTool,
+			UpgradeType:     upgrade.Spec.UpgradeType,
+			HistoryID:       newHistoryID,
+			Version:         upgrade.Spec.Version,
+			ImageRepository: upgrade.Spec.ImageRepository,
+		}
+
+		msg.BuildRouter(modules.UpgradeControllerModuleName, modules.UpgradeControllerModuleGroup, resource, "upgrade").
+			FillBody(upgradeReq)
+
+		err := dc.messageLayer.Send(*msg)
+		if err != nil {
+			klog.Errorf("Failed to send upgrade message %v due to error %v", msg, err)
```

Comment:
print `msg.ID` is better than print the complete msg. You can check the other logs as the same